doc/SEMANTICS.md

```diff
@@ -74,13 +74,27 @@ Synchronization operations (`fsync`, `fdatasync`) are currently no-ops because w
 
 Space allocation (`fallocate`, `posix_fallocate`) are not supported.
 
+#### Deletes
+
+File deletion (`unlink`) is planned but not yet supported (see [#78](https://github.com/awslabs/mountpoint-s3/issues/78)).
+When `unlink` is implemented, the following semantics are proposed.
+
+For files not yet committed to S3, the client will make a best effort to stop any inflight uploads,
+and immediately remove the file from its directory.
+If there are any open file handles to the file, future writes to them will fail.
```

Comment:
Alternatively, we can fail the unlink. It would fail faster since we don't have to cancel inflight uploads and probably less surprise?

Reply:
I hadn't considered this one.

Considering strict POSIX compliance, we should allow writing to continue (presumably also reading to continue if we supported that, with file handles able to see new content) after the `unlink`, and once these file handles are closed, the data is deleted from 'disk'.

We do want to try and be close to POSIX where feasible, so allowing `unlink` during write makes sense to me. Since we don't support reading files during writing, perhaps the better behavior to avoid surprise would be to allow writing to continue, with the bytes being discarded rather than the `write` operation returning EIO. However, I was concerned about one-way doors with this option and simply returning EIO is the simplest option for now. (Yes, allowing `unlink` to succeed is also a one-way door - but an intentional one.)

Reply:
FWIW, I am imagining this use case to be a very infrequent one. Thus, for customer applications that may try it (delete during write) we should fail in the way least likely to break applications and easy to understand what went wrong.

Reply:
Discussed offline - failing `unlink` for local files leaves open the path for implementing it closer to POSIX in future. We can implement this behavior as the initial case, and if there is demand we can extend the behavior.

Trying to handle the local files also presents complexity for now, so we can deliver `unlink` for remote files now and tackle local files later.

Thoughts @jamesbornholt?